### PR TITLE
Support MAVLINK_MSG_ID_SET_ATTITUDE_TARGET

### PR DESCRIPTION
**Pull request by lthall**
This PR supports the full Support MAVLINK_MSG_ID_SET_ATTITUDE_TARGET message as it is specified. It now supports
- quaternion attitude
- quaternion attitude + roll and/or pitch and/or yaw rate
- thrust or climb rate (climb rate is not to spec and must be turned off using GUID_OPTIONS)
- roll and/or pitch and/or yaw rate.

We need to look at how we support the old but incorrect behaviour of attitude roll/pitch + rate yaw.

I also need to add the getter message associated with the setter message.

Closes https://github.com/ArduPilot/ardupilot/issues/14610

### CODE REVIEW

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [inline ArduCopter/mode_guided.cpp:977]: Ahh I now see why you constructed a Quat(0,0,0,0). Might be nicer to just pass in attitude_ignore into here instead for those reading later.  

```suggestion
    if (ignore_attitude) {
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1073]: Yes, I agree. There are a number of issues here to discuss.

I used the 0,0,0,0 because that was something that has been defined in Mavlink here:
https://mavlink.io/en/messages/common.html#ATTITUDE_QUATERNION
"send [0, 0, 0, 0] if field not supported)"

Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1073]: > Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??

Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

File: libraries/AP_HAL_SITL/system.cpp
Comment (lthall) [reply libraries/AP_HAL_SITL/system.cpp:28]: This was actually a bug in master for Cygwin that I needed to patch here so I could compile. This will be fixed in master and will be able to be removed from here.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:215]: Yeh :)

File: ArduCopter/GCS_Mavlink.cpp
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp:1073]: > > Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??
> 
> Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

Yeah position_target_int is like that with passing so many variables. Possibly, the solution is to create a structure that holds all the needed variables for these functions and pass those into guided. Then we wouldn't need to change the function definitions which every change or additional feature.

**Comment by shubham-shahh** [comment]
Hi, I am using this PR with [Mavros_controllers](https://github.com/Jaeyoung-Lim/mavros_controllers), the UAV does random motion and then crashes. I would like to know if there's a difference in co-ordinate frames?

**Comment by lthall** [comment]
After talking to various Dev members I think we need to:

- Update guided logging to make logging clearer and maybe separate logging for position vs angle commands.
- Add and fix mavlink messages reporting the current state (for example current attitude target rather than the current attitude)
- Add messages to support angle + angle rate + climb rate (maybe)
- Add message to support  angle + angle rate limit + climb rate and/or throttle (maybe).

File: ArduCopter/version.h
Comment (amilcarlucas) [inline ArduCopter/version.h:9]: This is probably an unwanted change

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

**Comment by amilcarlucas** [comment]
Unfortunately this has merge conflicts now.

File: ArduCopter/mode_guided.cpp
Comment (lthall) [reply ArduCopter/mode_guided.cpp:981]: We don't have this variable.

File: ArduCopter/version.h
Comment (lthall) [reply ArduCopter/version.h]: Removed

File: ArduCopter/mode_guided.cpp
Comment (lthall) [reply ArduCopter/mode_guided.cpp:981]: We don't have this parameter. What we have now works so I would suggest we move forward with this and improve it in another PR.

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1112]: I would like to leave it like this and get it in and work on cleaning things up after.

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1100]: Done

**Comment by lthall** [comment]
I have rebased and I think I have addressed most of the problems. Given how much others are doing to further this I would like to get this done so I can get out of the way of others.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp
Comment (hendjoshsr71) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:242]:
```suggestion
    attitude_desired_update.from_axis_angle(ang_vel_target * _dt);
```

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.cpp
Comment (hendjoshsr71) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.cpp:149]:
```suggestion
void AC_AttitudeControl_Multi_6DoF::input_quaternion(const Quaternion& attitude_desired_quat, const Vector3f &ang_vel_target) {
```

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h
Comment (hendjoshsr71) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:23]:
```suggestion
    void input_quaternion(const Quaternion& attitude_desired_quat, const Vector3f &ang_vel_target) override;
```

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp
Comment (hendjoshsr71) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:214]:
```suggestion
void AC_AttitudeControl::input_quaternion(const Quaternion& attitude_desired_quaternion, const Vector3f &angular_vel_target)
{
    // Inputs
    Quaternion attitude_desired_quat = attitude_desired_quaternion;
    Vector3f ang_vel_target = angular_vel_target;
```

File: ArduCopter/GCS_Mavlink.cpp
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp:1112]: That's fine lets leave this open till it is handled by a later PR

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
None of these are have to but... I do think we should immediately follow up with checking the user's quaternion to be unit length. 

I haven't yet tested this again since your latest update, but here is a bit of an overview.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:214]: You are removing an important task of this function.

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:23]: I can do this for ang_vel_target but attitude_desired_quat isn't constant.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:242]: Yep, nice.

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1102]: changed to:
// ensure thrust field is not ignored

**Comment by lthall** [comment]
> Yeah ignore_attitude is up in where the type-mask in parsed in GCS_MAVLINK.
I'm ok with not fixing this for now. But lets leave the conversation open so I or someone else sees to come back to this.
I think it is important we check that the user's input quaternion meets our unit length expectation. But that can be done separately.


I think we can only ensure the quaternion is not zero as they are rarely exactly unit length. This is why we renormalise them regularly to ensure they don't move too far from unit length. We currently use the zero length to denote undefined so we have explicitly handled that case making normalisation safe after that.

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1112]: We still need to check for this to ensure we can normalise the quaternion safely.

File: ArduCopter/mode_guided.cpp
Comment (lthall) [reply ArduCopter/mode_guided.cpp:981]: > Yeah ignore_attitude is up in where the type-mask in parsed in GCS_MAVLINK.
I'm ok with not fixing this for now. But lets leave the conversation open so I or someone else sees to come back to this.
I think it is important we check that the user's input quaternion meets our unit length expectation. But that can be done separately.


I think we can only ensure the quaternion is not zero as they are rarely exactly unit length. This is why we renormalise them regularly to ensure they don't move too far from unit length. We currently use the zero length to denote undefined so we have explicitly handled that case making normalisation safe after that.

File: ArduCopter/GCS_Mavlink.cpp
Comment (lthall) [reply ArduCopter/GCS_Mavlink.cpp:1112]: It is also worth noting that we often use zero to denote a parameter is not defined.

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:23]: opps, no I can't. It is changed when limited.

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [reply ArduCopter/mode_guided.cpp:981]: Oh certainly. I didn't mean it had to be exactly unit length, but it should be close coming into the system. If is not "near ||q_vec|| = 1" the math could get weird fast as there are assumptions made that quaternions are near unit length.

File: ArduCopter/Copter.cpp
Comment (peterbarker) [inline ArduCopter/Copter.cpp:404]:
```suggestion
    mode_guided.set_angle(q, Vector3f{}, climb_rate_ms*100, false);
```

File: ArduCopter/GCS_Mavlink.cpp
Comment (peterbarker) [inline ArduCopter/GCS_Mavlink.cpp:1096]: I wonder if this code would read better if these were inverted so
```suggestion
        const bool use_roll_rate    = !(packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE);
```

... there's a bunch of double-negatives in this function that would disappear further down with this sort of change.

File: ArduCopter/mode_guided.cpp
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:981]:
```suggestion
    if (guided_angle_state.attitude_quat.is_zero()) {
```

new method needed

**Review by peterbarker — DISMISSED** [review]

File: ArduCopter/GCS_Mavlink.cpp
Comment (peterbarker) [inline ArduCopter/GCS_Mavlink.cpp:1109]:
```suggestion
            attitude_quat.zero();
```

File: ArduCopter/mode_guided.cpp
Comment (tridge) [reply ArduCopter/mode_guided.cpp:981]: let's add an is_zero() on Quaternion object

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/mode_guided.cpp
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:619]:
```suggestion
                           Vector3f{ToDeg(roll_rad), ToDeg(pitch_rad), ToDeg(pitch_rad)} * 100.0f,
```

Comment (peterbarker) [reply ArduCopter/mode_guided.cpp:619]: ... similarly elsewhere

File: ArduCopter/GCS_Mavlink.cpp
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp:1096]: This is the same pattern we have used right below this in `MAVLINK_MSG_ID_SET_POSITION_TARGET_LOCAL_NED` which  matches to what the mavlink type_mask is which is an ignore_mask not a use_mask.

If we want to change this then lets change it everywhere at once afterwards?

**Review by hendjoshsr71 — COMMENTED** [review]
I've updated this with the requested changes and done some basic testing using mavproxy's message module.

I've added as a separate commit a check to ensure the input quaternion is near the unit length assumption. 

@lthall you may want to review the changes?

File: ArduCopter/GCS_Mavlink.cpp
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp]: I think I handled this with the check for unit length input

**Review by tridge — APPROVED** [review]

**Review by hendjoshsr71 — APPROVED** [review]
LGTM

**Comment by hendjoshsr71** [comment]
This Closes https://github.com/ArduPilot/ardupilot/issues/14610

File: ArduCopter/mode.h
Comment (rmackay9) [inline ArduCopter/mode.h:875]: We should have a comment above this method and/or in the .cpp file clarifying that attitude_quat should be zero if only an angular velocity is provided.

Also the "const Quaternion attitude_quat" can be passed by reference.  So it should be, "const Quaternion& attitude_quat"

File: ArduCopter/mode_guided.cpp
Comment (rmackay9) [inline ArduCopter/mode_guided.cpp:983]: We don't normally leave commented out code around like this so I think this should be removed.

**Comment by rmackay9** [comment]
Sorry, I'm a little late to the party on this one.  I've found a couple of non-functional issues that could be fixed (see above) although that doesn't need to block the merge of this PR.

I've removed the "MergeOnCIPass" label but that's just so that @hendjoshsr71 sees my comments.  Maybe Josh will want to fix these before he merges them or maybe not.

In any case, I've given this a quick test with the copy-paste example from the MAVLink Interface page on the subject and confirmed that at least the existing functionality still works. https://ardupilot.org/dev/docs/copter-commands-in-guided-mode.html#copter-commands-in-guided-mode-set-attitude-target

File: ArduCopter/mode.h
Comment (hendjoshsr71) [reply ArduCopter/mode.h:875]: I don't think this is quite right. 

This new scheme allows you to specify both the attitude and rates at the same time. If the attitude quaternion is zero then we use only the angular velocity instead through a different function.

I've added a new comment to clear this up hoepfully?

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [reply ArduCopter/mode_guided.cpp:983]: Oh thanks! Missed this one!

File: libraries/AC_AttitudeControl/AC_AttitudeControl.h
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.h:297]: The comment mentions "smoothing_gain" but this argument has been renamed to "input_tc" now.

File: ArduCopter/mode.h
Comment (rmackay9) [reply ArduCopter/mode.h]: great, the comment looks very clear now, txs!

**Review by rmackay9 — APPROVED** [review]

**Comment by rmackay9** [comment]
This looks good to me.  I'm unsure why it isn't passing the autotests.  I've kicked them off again so if they pass we can merge.  If it doesn't pass then maybe this PR should be rebased on master with the 
hope that perhaps the flapping test has been resolved in master.

**Comment by rmackay9** [comment]
Still not passing so this probably needs a rebase on master.

**Comment by lthall** [comment]
That test is the:
  test.CopterTests1e:
    Mount (Test Camera/Antenna Mount) (Mount pitch not achieved) (see /tmp/buildlogs/ArduCopter-Mount.txt)

I was seeing this test fail before too. Does this test consume this message?

**Comment by hendjoshsr71** [comment]
I think i got the auto test failure fixed here. 

The autotest changes here removed the setting of angular rates as they were being used previously before this PR. Making these changes "equivalent" to previous behavior.

This autotest `test.Mount` could probably use some work too, but we shouldn't hold up this PR in my belief.

File: Tools/autotest/arducopter.py
Comment (amilcarlucas) [inline Tools/autotest/arducopter.py:4260]:
```suggestion
                0, # pitch rate (rad/s)
                0, # yaw rate (rad/s)
```

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

**Comment by hendjoshsr71** [comment]
Thank you @amilcarlucas for the review and the logging bug fix!

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [reply ArduCopter/mode_guided.cpp]: Fixed. Wow not how this got missed. Thanks!

File: ArduCopter/mode.h
Comment (hendjoshsr71) [reply ArduCopter/mode.h]: Fixed all of these thanks!

**Comment by IuliuNovac** [comment]
It looks like there is a bug.

If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.

So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
`    unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
`

**Comment by shubham-shahh** [comment]
> It looks like there is a bug.
> 
> If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.
> 
> So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
> `    unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
> `

Are you referring to the issue mentioned here? https://github.com/ArduPilot/ardupilot/issues/20444

**Comment by IuliuNovac** [comment]
> > It looks like there is a bug.
> > If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.
> > So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
> > `   unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;`
> 
> Are you referring to the issue mentioned here? #20444

Thanks, that seems to be right, but holly Christ, there so much emotions in that thread.

**Comment by lthall** [comment]
>> So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion. However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates. I am working with 4.4.2, but in master branch the last time i checked it was still the same.

The problem with this is the quaternion specifies all three axis. There is no way for this message to specify only roll and pitch angles because it's only attitude specification is the single axis quaternion rotation. When we ignore all rates what we are really saying is they are zero. This means the attitude we specify with the quaternion is the target stationary attitude. So when you specify a yaw rate you are really saying roll rate = 0, pitch rate = 0, yaw = something.

The reason for the emotion in the other threads is people want to specify the attitude in Euler angles with Euler rates but this message does not support Euler angles and rates at all. The reaction is to say, no problem I will just convert from Euler to quaternion and ignore the yaw. The problem is this message has no concept of Euler angles and therefore can never be used in this way.

I would like to support an easy Euler based attitude message that does support this input however I suspect that I don't get any real interest in this because everybody discovers they can get all they need and more once they work out how to use the current message correctly. The problem is there are a number of ways to use this message incorrectly that appear to be sensible at first glance.

**Comment by amilcarlucas** [comment]
@lthall your videos explain this in detail. It's just a pity that people do not search for the videos. If they would, they would be a lot more productive faster.

**Comment by IuliuNovac** [comment]
> @lthall your videos explain this in detail. It's just a pity that people do not search for the videos. If they would, they would be a lot more productive faster.

I am having a hard time finding this videos, could you suggest where to look for them?

**Comment by lthall** [comment]
Could you start a thread and repeat this question and the how the set attitude is supposed to be utilized question on the discus forums so we can talk there: https://discuss.ardupilot.org/

**Pull request merged**
